kytos/core/link.py

```diff
@@ -112,31 +114,32 @@ def is_tag_available(self, tag):
 
     def get_next_available_tag(self):
         """Return the next available tag if exists."""
-        # Copy the available tags because in case of error
-        # we will remove and add elements to the available_tags
-        available_tags_a = self.endpoint_a.available_tags.copy()
-        available_tags_b = self.endpoint_b.available_tags.copy()
-        random.shuffle(available_tags_a)
-        random.shuffle(available_tags_b)
-
-        for tag in available_tags_a:
-            # Tag does not exist in endpoint B. Try another tag.
-            if tag not in available_tags_b:
-                continue
-
-            # Tag already in use. Try another tag.
-            if not self.endpoint_a.use_tag(tag):
-                continue
-
-            # Tag already in use in B. Mark the tag as available again.
-            if not self.endpoint_b.use_tag(tag):
-                self.endpoint_a.make_tag_available(tag)
-                continue
-
-            # Tag used successfully by both endpoints. Returning.
-            return tag
-
-        raise KytosNoTagAvailableError(self)
+        with self._get_next_available_tag_lock:
+            # Copy the available tags because in case of error
+            # we will remove and add elements to the available_tags
+            available_tags_a = self.endpoint_a.available_tags.copy()
+            available_tags_b = self.endpoint_b.available_tags.copy()
+            random.shuffle(available_tags_a)
+            random.shuffle(available_tags_b)
+
+            for tag in available_tags_a:
+                # Tag does not exist in endpoint B. Try another tag.
+                if tag not in available_tags_b:
+                    continue
+
+                # Tag already in use. Try another tag.
+                if not self.endpoint_a.use_tag(tag):
+                    continue
+
+                # Tag already in use in B. Mark the tag as available again.
+                if not self.endpoint_b.use_tag(tag):
+                    self.endpoint_a.make_tag_available(tag)
+                    continue
+
+                # Tag used successfully by both endpoints. Returning.
+                return tag
+
+            raise KytosNoTagAvailableError(self)
 
     def make_tag_available(self, tag):
```

Comment:
@viniarck I'm working on fixing the other two possible race conditions related to `make_tag_available()` but I realize that mef_eline instantiates a Link object for each EVC and each link in the path. Thus, the Lock mechanism will not apply for inter-EVC coordination, which leaves the code still vulnerable to race conditions. I will have to change a little bit more this PR to be in compliance with this situation.

Reply:
@viniarck I've moved the Lock mechanism to Interface, because it is the object instantiated by Kytos core (and topology) that remains active for all dynamic Link objects created by mef_eline. Thus, they will protect the shared resource `Interface.available_tags`

Reply:
Sounds great to me, closest to core as possible. I was double checking here, @italovalcy , looks like we might also need the lock in the `is_tag_available`, that's also used by a [Link here - that's calling the underlying `Interface.is_tag_available`](https://github.com/kytos-ng/kytos/blob/master/kytos/core/link.py#L108-L111), otherwise we might not have an atomic read of the underlying `self.tags_available` in that part and might result in issues? Could you also double check this part? I'll leave PR approved. Thanks.

Reply:
Thanks for the double check, @viniarck. I've addresses your concern on commit 42c0433. I ask you please to recheck.

Reply:
Looks good to me. Thanks, @italovalcy.